File: src/main/java/nz/co/searchwellington/controllers/RssController.java
package nz.co.searchwellington.controllers;

import java.util.Map;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import nz.co.searchwellington.model.SiteInformation;
import nz.co.searchwellington.repositories.ContentRetrievalService;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;

import uk.co.eelpieconsulting.common.views.ViewFactory;
import uk.co.eelpieconsulting.common.views.rss.RssView;

import com.google.common.collect.Maps;

@Controller	// TODO doesn't need to be a special case?
public class RssController {
	
    private SiteInformation siteInformation;
	private ContentRetrievalService contentRetrievalService;
	private ViewFactory viewFactory;
	
	@Autowired
    public RssController(SiteInformation siteInformation, ContentRetrievalService contentRetrievalService, RssUrlBuilder rssUrlBuilder, ViewFactory viewFactory) {
        this.siteInformation = siteInformation;
        this.contentRetrievalService = contentRetrievalService;
		this.viewFactory = viewFactory;
    }
    
    @RequestMapping("/rss")
    public ModelAndView mainRss(HttpServletRequest request, HttpServletResponse response) throws Exception {    	
		final Map<String, Object> model = Maps.newHashMap();		
		String title = siteInformation.getAreaname() + " Newslog";
		String link = siteInformation.getUrl();
        String description = "Links to " + siteInformation.getAreaname() + " related newsitems.";
        model.put("data", contentRetrievalService.getLatestNewsitems());
        
        final RssView rssView = viewFactory.getRssView(title, link, description);
        return new ModelAndView(rssView, model);        
    }
    
}

File: src/main/scala/nz/co/searchwellington/controllers/RssController.scala
package nz.co.searchwellington.controllers

import java.util

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import com.google.common.collect.Maps
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model.SiteInformation
import nz.co.searchwellington.repositories.ContentRetrievalService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.servlet.ModelAndView
import uk.co.eelpieconsulting.common.views.ViewFactory

import scala.concurrent.Await

@Controller class RssController @Autowired()(siteInformation: SiteInformation, contentRetrievalService: ContentRetrievalService,
                                             rssUrlBuilder: RssUrlBuilder, viewFactory: ViewFactory) extends ReasonableWaits {

  @RequestMapping(Array("/rss")) // TODO Should be a special case of the index model builder.
  @throws(classOf[Exception])
  def mainRss(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    val title = siteInformation.getAreaname + " Newslog"
    val link = siteInformation.getUrl
    val description = "Links to " + siteInformation.getAreaname + " related newsitems."

    val model: util.HashMap[String, Any] = Maps.newHashMap()
    import scala.collection.JavaConverters._
    model.put("data", Await.result(contentRetrievalService.getLatestNewsitems(), TenSeconds).asJava)
    new ModelAndView(viewFactory.getRssView(title, link, description), model)
  }

}
